### DISCUSSION TITLE
Would be it possible to build libraries as `dialog`, `gtk4`, `libadwaita`, `libqalculate` and `yad` in AppImage format to work on the corporation version of Red hat Linux?

### DISCUSSION THREAD

**Discussion by gusbemacbe** [AppImage/AppImageKit#1263] [Q&A] [answered]
Hello!

The coordinator of the team, who takes care of the Linux images, said that they do not want to approve the software they find unnecessary for the job/work, keeping the job/operating system lean and optimised. 

The solution is to use AppImages. 

But `dialog`, `gtk4`, `libadwaita`, `libqalculate` and `yad` are not software. They are libraries. Of course that I have seen the case of `ImageMagick`. But if I build `gtk4` and `libadwaita` in AppImage format, will the GNOME extensions like @jmmaranan's [Forge](https://github.com/jmmaranan/forge) and @christopher-l's [Space Bar](https://github.com/christopher-l/space-bar) work with these libraries in AppImage format?

**Comment by gusbemacbe** [comment]
Hello @probonopd and @TheAssassin!

I was able to build all these libraries.

I tested on Red Hat Linux and all they failed due to countless missed libraries. I will have to include all the missed libraries. 

But I am worried. Would it be possible to isolate the libraries, for example, `libc.so`, for not overwriting Red Hat Linux's `libc.so`, so the application will match only its own `libc.so` and will not touch Red Hat Linux's `libc.so`; Red Hat Linux will not touch the application's `libc.so`? 

You can get my compressed file: https://github.com/gusbemacbe/dotfiles-minimos/releases/download/v10.0/AppImages.tar.gz

**Comment by gusbemacbe** [comment]
I tested with `linuxdeploy` and it didn't help either.

**Comment by probonopd** [comment]
Hello @gusbemacbe, which applications would you like to see as AppImages?

Please ask the authors of the applications you would like to use to provide these applications in AppImage format.

**Comment by gusbemacbe** [comment]
Hello @probonopd!

> Hello @gusbemacbe, which applications would you like to see as AppImages?

Only `gtk4`, `libadwaita` and `yad` are not applications, but are dependencies for two GNOME extensions to work. And I also need `dialog` and `libqalculate`, that are really not applications. They are like ImageMagick.

> Please ask the authors of the applications you would like to use to provide these applications in AppImage format.

The authors of `gtk4` and `libadwaita` are GNOME developers. I doubt that they be interested in developing them in AppImage format. Some of them do not have GitHub account. Their repositories: 

- https://gitlab.gnome.org/GNOME/gtk
- https://gitlab.gnome.org/GNOME/libadwaita

The authors of `dialog` do not have a Github account and did not host their repository on GitHub. His repository is https://invisible-island.net/dialog/. 

Evangelos (@foutrelis) takes care of this repository for Arch Linux. I am uncertain that he can help to build `dialog` in AppImage format. 

The collaborators of `libqalculate` host their repository on GitHub. I will mark one company and some of them:

- @Qalculate
- @alexhenrie, @chengsun, @Daninet, @hanna-kn, @idamir, @NorwayFun, and @Polakrity;

The author of `yad` is @v1cont. I am uncertain that he can help because he has difficulties as he is in Ukraine amid the war.

**Comment by probonopd** [comment]
As the name suggests, AppImage is for applications.

It cannot be used to "just" ship libraries for use by Gnome extensions.

But if you want to ship an application that uses e.g., libadwaita, then you can bundle libadwaita with the application that uses it in the AppImage.

**Comment by v1cont** [comment]
Hi

I'm an author of YAD. There is a modern war here, so we have a good internet link even in
Bakhmut trenches (big thans to Elon Musk) and some free time (huge thanks to our artilery).
So periodically i can do what i like to do. But i'm not sure what kind of help you are asked 
from me?

Fri, 14 Apr 2023 12:11:09 -0700
Gustavo Reis ***@***.***> написав:

> Hello @probonopd!
> 
> > Hello @gusbemacbe, which applications would you like to see as AppImages?
> 
> Only `gtk4`, `libadwaita` and `yad` are not applications, but are dependencies for
> two GNOME extensions to work. And I also need `dialog` and `libqalculate`, that
> are really not applications. They are like ImageMagick.
> 
> > Please ask the authors of the applications you would like to use to provide
> > these applications in AppImage format.
> 
> The authors of `gtk4` and `libadwaita` are GNOME developers. I doubt that they be
> interested in developing them in AppImage format. Some of them do not have GitHub
> account. Their repositories: 
> 
> - https://gitlab.gnome.org/GNOME/gtk
> - https://gitlab.gnome.org/GNOME/libadwaita
> 
> The authors of `dialog` do not have a Github account and did not host their
> repository on GitHub. His repository is https://invisible-island.net/dialog/. 
> 
> Evangelos ***@***.***) takes care of this repository for Arch Linux. I am
> uncertain that he can help to build `dialog` in AppImage format. 
> 
> The collaborators of `libqalculate` host their repository on GitHub. I will mark
> one company and some of them:
> 
> - @Qalculate
> - @alexhenrie, @chengsun, @Daninet, @hanna-kn, @idamir, @NorwayFun, and @Polakrity;
> 
> The author of `yad` is @v1cont. I am uncertain that he can help because he has
> difficulties as he is in Ukraine amid the war. 
> 
> 


-- 
wbr
Victor Ananjevsky ***@***.***>

**Comment by gusbemacbe** [comment]
@probonopd 

> As the name suggests, AppImage is for applications.

`dialog`, `libqalculate` and `yad` are like ImageMagick. The ImageMagick developers were able to build an AppImage for ImageMagick, and I could run `magick` from ImageMagick in AppImage.

The problem is with `dialog`, `libqalculate` and `yad` are the missed dependencies. I can not update my company's corporative version of Red Hat Linux 8.7. Red Hat Linux 8.7 uses the old versions of dependencies/libraries, most of them are dated in 2015 and 2016. Therefore, I need to ship all dependencies/libraries to bundle with these binaries, but I also need to isolate the new version of `glibc`, that these binaries require, from Red Hat Linux, that uses the old version. 

I will show the screenshot to @v1cont, author of `yad`. Check below.

---

Pryvit @v1cont!

I hope you be good,  as I also hope that @SmartFinn, author of Papirus icons theme, is good. 

> So periodically i can do what i like to do. But i'm not sure what kind of help you are asked
from me?

I would like to build your binary, but bundling or shipping all the missed dependencies required by your binary. Red Hat Linux is a Jurassic distribution. 

Let me show the screenshots I took from my company's computer with Red Hat Linux 8.7:

![image](https://user-images.githubusercontent.com/2257024/232256484-c88e51ab-929d-4e06-8207-1cb6e76bd069.png)
![image](https://user-images.githubusercontent.com/2257024/232256514-e339867d-3281-47d5-984f-c6ddf44ac989.png)

**Comment by gusbemacbe** [comment]
I installed Red Hat Linux Enterprise 8.7 on VirtualBox, similar to my company's computer's SO. I checked the list of installed packages, then it seems that these binaries do not support the version 2.28 of `glibc`. 

```
glib-networking.x86_64                             2.56.1-1.1.el8                                    @rhel-8-for-x86_64-baseos-rpms   
glib2.x86_64                                       2.56.4-159.el8                                    @rhel-8-for-x86_64-baseos-rpms   
glibc.x86_64                                       2.28-211.el8                                      @rhel-8-for-x86_64-baseos-rpms   
glibc-all-langpacks.x86_64                         2.28-211.el8                                      @rhel-8-for-x86_64-baseos-rpms   
glibc-common.x86_64                                2.28-211.el8                                      @rhel-8-for-x86_64-baseos-rpms   
glibc-devel.x86_64                                 2.28-211.el8                                      @rhel-8-for-x86_64-baseos-rpms   
glibc-gconv-extra.x86_64                           2.28-211.el8                                      @rhel-8-for-x86_64-baseos-rpms   
glibc-headers.x86_64                               2.28-211.el8                                      @rhel-8-for-x86_64-baseos-rpms   
glibc-langpack-en.x86_64                           2.28-211.el8                                      @rhel-8-for-x86_64-baseos-rpms   
```

I am uncertain if this version of GNOME is compatible with these two extensions:

```
gnome-shell.x86_64                                 3.32.2-48.el8                                     @rhel-8-for-x86_64-appstream-rpms
gnome-shell-extension-apps-menu.noarch             3.32.1-29.el8                                     @rhel-8-for-x86_64-appstream-rpms
gnome-shell-extension-common.noarch                3.32.1-29.el8                                     @rhel-8-for-x86_64-appstream-rpms
gnome-shell-extension-desktop-icons.noarch         3.32.1-29.el8                                     @rhel-8-for-x86_64-appstream-rpms
gnome-shell-extension-horizontal-workspaces.noarch 3.32.1-29.el8                                     @rhel-8-for-x86_64-appstream-rpms
gnome-shell-extension-launch-new-instance.noarch   3.32.1-29.el8                                     @rhel-8-for-x86_64-appstream-rpms
gnome-shell-extension-places-menu.noarch           3.32.1-29.el8                                     @rhel-8-for-x86_64-appstream-rpms
gnome-shell-extension-window-list.noarch           3.32.1-29.el8                                     @rhel-8-for-x86_64-appstream-rpms
```

I noticed that Red Hat Linux Enterprise 8.7's repositories do not have `gtk4`, `libadwaita` and `yad`. 

Is there how to build the latest version of GNOME Shell, bundling all the dependencies and libraries, in AppImage format, for Red Hat Linux?

I know how to build `libqalculate` and `yad` in AppImage format, but I need to know which dependencies and libraries these binaries depend on and require, to bundle. I need to know to isolate AppImage application's libraries from Red Hat.

**Comment by probonopd** [comment]
So what you really want to do is to make a `yad.AppImage` that either is built on el8 or older, or bundles _all_ dependencies includig glibc.

The latter can be done using the `-s deploy` option of https://github.com/probonopd/go-appimage/tree/master/src/appimagetool.

**Comment by gusbemacbe** [comment]
@probonopd, 

I have the folder called `yad`. I ran the command `appimagetool -s yad`. I tested on Red Hat Linux, and it failed. It appears that this command did not bundle the dependencies of the base system. Was my command wrong?

**Reply by probonopd** [reply]
Well, you need to use it like in the README:

`./appimagetool-*.AppImage -s deploy yad/usr/share/applications/yad.desktop`

And of course you need a ` yad/usr/...` folder structure with `yad/usr/share/applications/yad.desktop` at the correct location.

**Reply by gusbemacbe** [reply]
It worked perfectly on Red Hat, but when I symlink `yad.AppImage` to `yad` in the directory `~/.local/bin`. Then I ran the command `yad --help`, and it did not tell the version, showed `yad-icon-browser` details and options instead. 

I have adapted the `AppRun` file, adding this code:

```bash
# Original
HERE="$(dirname "$(readlink -f "${0}")")"

cd "$HERE/usr" # Not all applications will need this; TODO: Make this opt-in
MAIN_BIN=$(find "$HERE/usr/bin" -name "$MAIN" | head -n 1)

# Added
if [ -n "$APPIMAGE" ] ; then
  BINARY_NAME=$(basename "$ARGV0")
  if [ -e "$HERE/bin/$BINARY_NAME" ] ; then
    exec "$HERE/bin/$BINARY_NAME" "$@"
  else
    exec "$HERE/bin/yad" "$@"
    exec "$HERE/bin/yad-icon-browser" "$@"
    exec "$HERE/bin/yad-tools" "$@"
  fi
else
    exec "$HERE/bin/yad" "$@"
    exec "$HERE/bin/yad-icon-browser" "$@"
    exec "$HERE/bin/yad-tools" "$@"
fi
```

But it still showed the `yad-icon-browser` details and options.

**Reply by probonopd** [reply]
Where is `$MAIN` supposed to come from? Can you upload your AppImage somehere for testing?

**Reply by gusbemacbe** [reply]
> Where is `$MAIN` supposed to come from?

It came from the command that you gave me. It itself generated the `AppRun` file. The `appimagetool` version is 119. 

Follow the zipped file:

https://github.com/gusbemacbe/dotfiles-minimos/releases/download/v10.0/yad.tar.gz

**Reply by probonopd** [reply, marked as answer]
Thanks. Look where your app is running - on FreeBSD :-)

![image](https://user-images.githubusercontent.com/2480569/232309705-7eebe077-dac9-4bf5-9a26-5efc5300be13.png)

This shows the advantage of AppImages that bundle _everything_. :100: 

As for your question, it seems like your code never gets executed, because the code block directly above your addition runs the application.

You would need to modify the existing block instead of appending your own. Like this:

```
############################################################################################
# Run experimental bundle that bundles everything if a private ld-linux-x86-64.so.2 is there
# This allows the bundle to run even on older systems than the one it was built on
############################################################################################

cd "$HERE/usr" # Not all applications will need this; TODO: Make this opt-in

# Try to find a binary with the same name as the AppImage or the symlink through which
# it was invoked, without any suffix
env
if [ -z "$ARGV0" ] ; then
  # AppRun is being executed outside of an AppImage
  ARGV0="$0"
fi
BINARY_NAME=$(basename "$ARGV0")
BINARY_NAME="${BINARY_NAME%.*}" # remove everything after the last "."
MAIN_BIN=$(find "$HERE/usr/bin" -name "$BINARY_NAME" | head -n 1)

# Fall back to finding the main binary based on the Exec= line in the desktop file
if [ -z "$BINARY_NAME" ] ; then
  MAIN_BIN=$(find "$HERE/usr/bin" -name "$MAIN" | head -n 1)
fi

LD_LINUX=$(find "$HERE" -name 'ld-*.so.*' | head -n 1)
...
```

You get the idea. I haven't fully tested this, so let me know how this works for you. Test it inside and outside the AppImage (that is, by symlinking and running `AppRun` directly).

If it works as intended, I'd like to make this the default for everyone.

**Reply by gusbemacbe** [reply]
I removed the part:

```shell
cd "$HERE/usr" # Not all applications will need this; TODO: Make this opt-in
MAIN_BIN=$(find "$HERE/usr/bin" -name "$MAIN" | head -n 1)
LD_LINUX=$(find "$HERE" -name 'ld-*.so.*' | head -n 1)
if [ -e "$LD_LINUX" ] ; then
  export GCONV_PATH="$HERE/usr/lib/gconv"
  export FONTCONFIG_FILE="$HERE/etc/fonts/fonts.conf"
  export GTK_EXE_PREFIX="$HERE/usr"
  export GTK_THEME=Default # This one should be bundled so that it can work on systems without Gtk
  export GDK_PIXBUF_MODULEDIR=$(find "$HERE" -name loaders -type d -path '*gdk-pixbuf*')
  export GDK_PIXBUF_MODULE_FILE=$(find "$HERE" -name loaders.cache -type f -path '*gdk-pixbuf*') # Patched to contain no paths
  # export LIBRARY_PATH=$GDK_PIXBUF_MODULEDIR # Otherwise getting "Unable to load image-loading module"
  export XDG_DATA_DIRS="${HERE}"/usr/share/:"${XDG_DATA_DIRS}"
  export PERLLIB="${HERE}"/usr/share/perl5/:"${HERE}"/usr/lib/perl5/:"${PERLLIB}"
  export GSETTINGS_SCHEMA_DIR="${HERE}"/usr/share/glib-2.0/runtime-schemas/:"${HERE}"/usr/share/glib-2.0/schemas/:"${GSETTINGS_SCHEMA_DIR}"
  export QT_PLUGIN_PATH="${HERE}"/usr/lib/qt4/plugins/:"${HERE}"/usr/lib/i386-linux-gnu/qt4/plugins/:"${HERE}"/usr/lib/x86_64-linux-gnu/qt4/plugins/:"${HERE}"/usr/lib32/qt4/plugins/:"${HERE}"/usr/lib64/qt4/plugins/:"${HERE}"/usr/lib/qt5/plugins/:"${HERE}"/usr/lib/i386-linux-gnu/qt5/plugins/:"${HERE}"/usr/lib/x86_64-linux-gnu/qt5/plugins/:"${HERE}"/usr/lib32/qt5/plugins/:"${HERE}"/usr/lib64/qt5/plugins/:"${QT_PLUGIN_PATH}"
  # exec "${LD_LINUX}" --inhibit-cache --library-path "${LIBRARY_PATH}" "${MAIN_BIN}" "$@"
  case $line in
    "ld-linux"*) 
      exec "${LD_LINUX}" --inhibit-cache "${MAIN_BIN}" "$@"
      exec "${LD_LINUX}" --inhibit-cache yad "$@"
      exec "${LD_LINUX}" --inhibit-cache yad-icon-browser "$@"
      exec "${LD_LINUX}" --inhibit-cache yad-tools "$@"
      ;;
    *) 
      exec "${LD_LINUX}" "${MAIN_BIN}" "$@" 
      exec "${LD_LINUX}" yad "$@" 
      exec "${LD_LINUX}" yad-icon-browser "$@" 
      exec "${LD_LINUX}" yad-tools "$@"
      ;;
  esac
else
  exec "${MAIN_BIN}" "$@"
  exec yad "$@" 
  exec yad-icon-browser "$@" 
  exec yad-tools "$@"
fi
```

I copied this code you gave me. After the test, it shows the complete list of environment variables instead.

Maybe you should give a look at ImageMagick's `AppRun`:

```shell
HERE="$(dirname "$(readlink -f "${0}")")"

export MAGICK_HOME="$HERE/usr:$MAGICK_HOME" # https://imagemagick.org/QuickStart.txt
export MAGICK_CONFIGURE_PATH=$(readlink -f "$HERE/usr/lib/ImageMagick-7.0.9/config-Q16"):$(readlink -f "$HERE/usr/lib/ImageMagick-7.0.9/config-Q16HDRI"):$(readlink -f "$HERE/usr/share/ImageMagick-7"):$(readlink -f "$HERE/usr/etc/ImageMagick-7"):$MAGICK_CONFIGURE_PATH #Wildcards don't work

export LD_LIBRARY_PATH=$(readlink -f "$HERE/usr/lib"):$LD_LIBRARY_PATH
export LD_LIBRARY_PATH=${HERE}/usr/lib/ImageMagick-7.0.9/modules-Q16HDRI/coders:$LD_LIBRARY_PATH

if [ "$1" == "man" ] ; then
  export MANPATH="$HERE/usr/share/man:$MANPATH" ; exec "$@" ; exit $?
elif [ "$1" == "info" ] ; then
  export INFOPATH="$HERE/usr/share/info:$INFOPATH" ; exec "$@" ; exit $?
fi

if [ -n "$APPIMAGE" ] ; then
  BINARY_NAME=$(basename "$ARGV0")
  if [ -e "$HERE/usr/bin/$BINARY_NAME" ] ; then
    exec "$HERE/usr/bin/$BINARY_NAME" "$@"
  else
    exec "$HERE/usr/bin/magick" "$@"
  fi
else
  exec "$HERE/usr/bin/magick" "$@"
fi
```

And I linked like:

```shell
ln -sfnr ImageMagick.Appimage ~/.local/bin/convert
ln -sfnr ImageMagick.Appimage ~/.local/bin/magick
ln -sfnr ImageMagick.Appimage ~/.local/bin/mgorify
```

When I ran `convert`, `magick` and `mogrify`, they showed the different and correct details. 

Perhaps speak with the ImageMagick developers about how they managed to do this:

https://github.com/ImageMagick/ImageMagick

And who wrote the `AppRun` file:

* @dlemstra
* @ruanformigoni

**Reply by probonopd** [reply]
> I copied this code you gave me. After the test, it shows the complete list of environment variables instead.

That is because I added a line `env` during development.

Does my code not work for you? Then you may need to edit a little bit. No need to look elsewhere.

Of course you must NOT delete everything after `LD_LINUX=$(find "$HERE" -name 'ld-*.so.*' | head -n 1)` from the original code. That is what the `...` means in the code I gave you.

**Reply by gusbemacbe** [reply]
@probonopd,

I followed your last instruction, and it worked, but only the command `yad --help` showed the correct details. The `yad-icon-browser` and `yad-tools` showed the wrong details because they showed the same details of `yad`. 

Follow my last version of AppImage with a `test` folder for you to test running the commands: https://github.com/gusbemacbe/dotfiles-minimos/releases/download/v10.0/yad.tar.gz

- `test/yad --help`
- `test/yad-icon-browser --help`
- `test/yad-tools --help`

**Reply by probonopd** [reply]
Also when you use the unpacked AppImage's AppRun file directly? That's what I had tested and what seemed to have worked for me.

**Reply by gusbemacbe** [reply]
Now it worked with your code, I figured out that I had to remove two line in red (see below):

```diff
cd "$HERE/usr" # Not all applications will need this; TODO: Make this opt-in

# Try to find a binary with the same name as the AppImage or the symlink through which
# it was invoked, without any suffix
env
if [ -z "$ARGV0" ] ; then
  # AppRun is being executed outside of an AppImage
  ARGV0="$0"
fi
BINARY_NAME=$(basename "$ARGV0")
BINARY_NAME="${BINARY_NAME%.*}" # remove everything after the last "."
MAIN_BIN=$(find "$HERE/usr/bin" -name "$BINARY_NAME" | head -n 1)

# Fall back to finding the main binary based on the Exec= line in the desktop file
if [ -z "$BINARY_NAME" ] ; then
  MAIN_BIN=$(find "$HERE/usr/bin" -name "$MAIN" | head -n 1)
fi

LD_LINUX=$(find "$HERE" -name 'ld-*.so.*' | head -n 1)

- MAIN_BIN=$(find "$HERE/usr/bin" -name "$MAIN" | head -n 1)
- LD_LINUX=$(find "$HERE" -name 'ld-*.so.*' | head -n 1)
```

The commands `yad`, `yad-icon-browser` and `yad-tools` showed the correct details. I will mark your answer with that code as solved.

**Reply by gusbemacbe** [reply]
@probonopd 

* `dialog` works perfectly. 
* `libqalculate` almost works, but has few failures:
    
    ![image](https://user-images.githubusercontent.com/2257024/233175901-2eeefbfb-e8a6-47d4-9356-2ff8b1ca3537.png)

    You link `qalculate.AppImage` to `qalc`, then enter `./qalc`. You can get my package and analyse: https://github.com/gusbemacbe/dotfiles-minimos/releases/download/v10.0/qalculate.tar.gz

**Reply by probonopd** [reply]
Sure, you need to make it so that it doesn't attempt to load stuff from paths at `/usr/`, because the files are in the AppDir/AppImage, not at `/usr/`. You need to change qalculate so that it loads its resources from a path relative to itself, rather than from an absolute path. (This has nothing to do with the `AppRun` script but is an entirely different subject - making the application/library "relocatable in the filesystem" so that it can be freely moved around.)

Check 
https://doc.qt.io/qt-5/qcoreapplication.html#applicationDirPath
This requires you to make code-level changes, and to recompile the code.

As a quick and dirty workaround, you might get away by replacing `/usr` with `././` in your binaries instead.
For this, you don't need to recompile the code:

```
find your.AppDir/ -type f -exec sed -i '' -e 's|/usr|././|g' {} \;
```

**Reply by gusbemacbe** [reply]
It still failed. It output the errors: `$HERE././/bin`, etc. and they said they did not find them.

**Reply by probonopd** [reply]
You must not patch the AppRun file! Only the binary ingredients.

So, on an unpatched copy, run

```
find your.AppDir/usr -type f -exec sed -i '' -e 's|/usr|././|g' {} \;
```

to patch only what is inside `your.AppDir/usr`. (if this is where your binary ingredients are).

Sorry if this was not clear.

Again, this is a somewhat hackish workaround and may require some experimentation; the proper solution is to change the source code so that no absolute paths are hardcoded.

**Reply by probonopd** [reply]
I have modified the script so that everyone will get the symlink/rename functionality for free:

https://github.com/probonopd/go-appimage/commit/44951e8612feef500d9c6aa37a13b20ce484ef58 (Search for main binary based on the name of the AppDir, AppRun, or their symlink)

Now, to fix your https://github.com/gusbemacbe/dotfiles-minimos/releases/download/v10.0/qalculate.tar.gz:

```
# Extract
./qalculate.AppImage --appimage-extract

# Fix permissions, otherwise it won't run from the AppDir
chmod +x ./squashfs-root/usr/lib64/ld-linux-x86-64.so.2

# Patch absolute paths away
find squashfs-root/usr -type f -exec sed -i '' -e 's|/usr|././|g' {} \;

# Test
./squashfs-root/AppRun

# Test as a symlink
ln -sf ./squashfs-root/AppRun ./qalc
./qalc

# Repackage as an AppImage
VERSION=0 ./mkappimage-*-x86_64.AppImage squashfs-root/

# Test
./Qalculate\!-0-x86_64.AppImage

# Test as a symlink
ln -sf Qalculate\!-0-x86_64.AppImage qalc 
./qalc 
```

Note to self: On AlmaLinux (which is like CentOS but with a working Live ISO), possibly we need 

```
unset which_declare
unset which
```

RH-like distributions seem to be doing strange things in /etc/profile.d/which2.sh potentially causing breakage.

**Comment by ruanformigoni** [comment]
> 1. Keeping the job/operating system lean and optimised
> 1. I can not update my company's corporative version of Red Hat Linux 8.7. Red Hat Linux 8.7 uses the old versions of dependencies/libraries.
> 1. Would it be possible to isolate the libraries, for example, libc.so, for not overwriting Red Hat Linux's libc.so, so the application will match only its own libc.so and will not touch Red Hat Linux's libc.so

From the points you stated above I think you need a rootless container, not an AppImage. I've created a tool, [arts](https://gitlab.com/formigoni/arts), that should meet these points:

```sh
export ARTS_RW=1
./alpine.arts arts-resize 10G
./alpine.arts arts-root apk add xfce4 yad
./alpine.arts arts-cmd GDK_BACKEND=x11 yad
mv alpine.arts yad
```

Now you have a portable `yad` binary that uses the container's `libc.so` [3] and does not have access to the host's root filesystem (and therefore does not use host's libraries [2]).

Have you checked if you can use docker with your system administrator? You can run GUI programs in it as well. Another option is to use appimagebuilder to take care of dependencies for you.

**Reply by gusbemacbe** [reply]
It failed. Do I have to run it inside a Docker container?

![image](https://user-images.githubusercontent.com/2257024/233049647-ef1aee07-a197-4d91-b553-a8f8bc9de3a7.png)

**Reply by ruanformigoni** [reply]
Not really, the issue is probably that glibc is too old to run the container itself (oldest I've tried was debian 11's). We can continue the discussion [here](https://github.com/ruanformigoni/arts/discussions) if you'd like, to not go off-topic.

**Reply by ruanformigoni** [reply]
It worked fine for me on a rhel 8.7 VM:

![screenshot-167](https://user-images.githubusercontent.com/18180693/233073656-109dc5ac-9012-4232-bf72-5181bcd05bcd.png)

![screenshot-168](https://user-images.githubusercontent.com/18180693/233074327-e6987969-b736-42bd-9f7a-e4c9a4b3fe72.png)

**Reply by gusbemacbe** [reply]
Interesting, did you test the commands like:
```shell
./yad --help
./yad-icon-browser --help
./yad-tools --help
```
?

**Reply by ruanformigoni** [reply]
Sure:

![screenshot-170](https://user-images.githubusercontent.com/18180693/233109091-fd643586-3bb4-434e-99e9-9c74f33ac294.png)
![screenshot-169](https://user-images.githubusercontent.com/18180693/233109098-d7744f12-0afd-4c7e-bcf1-9b5cdd2cee9e.png)
